docs/Developers/NextGen_Core.rst

```diff
@@ -0,0 +1,286 @@
+``result_callback``, ``cache`` and ``executor`` (the last two are ``None`` by
+default) and initiates a coala session.
+
+* The ``bears`` argument contains the list of bears to be run for a coala
+  session.
+* The ``result_callback`` is a function that is called when the results from a
+  bear run are available.
+* The ``cache`` argument if provided enables caching and runs the session using
+  the cache provided to store the bear results.
+* The ``executor`` argument is used to provide a custom executor in which the
+  passed bears are to be run which is closed after the core is closed. If this
+  argument is not provided then ``ProcessPoolExecutor`` is used using as many
+  processes as cores available on the system.
+
+Official support for virtual files
+----------------------------------
+
+IDEs like IntelliJ use a virtual files to represent files in a filesystem (VFS)
+and perform operations on them. Hence NextGen-Core provides official support
+for virtual files, the bears have to point to the right file data objects when
+run, whether they are real files or virtual ones. This makes coala easier to
+integrate with IDEs since the contents reside in the IDE memory which is faster
+instead of reloading files from the disk.
+
+Task Objects
+------------
+
+These are the representation of the tasks performed by the bears. Structure-wise
+they are a tuple containing the tuple of positional arguments and a dict of
+keyword arguments to the bears. The task objects are instrumental to the caching
+mechanism as their hash values are stored in the cache along with the bear
+results and during each coala run the core looks for identical task objects in
+the cache, and fetches their corresponding results if found without even
+executing those tasks thereby providing a performance boost. Below you can see
+an example of two task objects.
+
+::
+
+    [
+        [(file, filename), {'filename_prefix': "", 'filename_suffix': ""}],
+        [(filename, file), {'preferred_quotation': ""}],
+    ]
+
+Improved dependency system
+--------------------------
+
+The NextGen-Core also introduces a better dependency management system than the
+one used by the old core. It features the following improvements:
+
+* A class ``DependencyTracker`` manages this task. The dependencies
+  can be added and resolved by this class and it also checks for circular
+  dependencies.
+* The dependency relations between two objects are tracked using a directed
+  graph. When two objects are added to this graph as an edge they form a
+  dependency relation.
+* Bears don't have to pass even results anymore.
+
+DependencyBear
+--------------
+
+The handling of bear dependencies by the old core wasn't effective. The
+old core used a queuing mechanism to communicate between bear runs which
+led to problems of a GlobalBear not running if any one of its
+dependencies (a local bear) terminated, the NextGen-Core improves on this.
```

Comment:
Have to reread this again, but one main point I'm just seeing that makes actually a difference to the old core is that the old core passed dependencies for all bears inside the `run`-argument `dependency_results`. Currently only the `DependencyBear` is recommended to be used for dependency handling to support caching.